components/ContainedInputViews/src/MDCContainedInputView.h

```diff
@@ -0,0 +1,319 @@
+ */
+@property(nonatomic, assign, readonly) MDCContainedInputViewFloatingLabelState floatingLabelState;
+
+/**
+ The @c floatingLabel is a label that occupies the text area when there is no text and that floats
+ above the text once there is some. It is distinct from a placeholder.
+ */
+@property(strong, nonatomic, readonly, nonnull) UILabel *floatingLabel;
+/**
+ The @c leadingUnderlineLabel can be used to display helper or error text.
+ */
+@property(strong, nonatomic, readonly, nonnull) UILabel *leadingUnderlineLabel;
+
+/**
+ The @c trailingUnderlineLabel can be used to display helper or error text.
+ */
+@property(strong, nonatomic, readonly, nonnull) UILabel *trailingUnderlineLabel;
+
+/**
+ This property is used to determine how much horizontal space to allot for each of the two underline
+ labels.
+
+ @note The default value is MDCContainedInputViewUnderlineLabelDrawPriorityTrailing. The rationale
+ behind this is it is less likely to have long explanatory error text and more likely to have short
+ text, like a character counter. It is better to draw the short text first and use whatever space is
+ leftover for the longer text, which may wrap to new lines.
+ */
+@property(nonatomic, assign)
+    MDCContainedInputViewUnderlineLabelDrawPriority underlineLabelDrawPriority;
+
+/**
+ When @c underlineLabelDrawPriority is set to @c .custom the value of this property helps determine
+ what percentage of the available width each underline label gets. It can be thought of as a
+ divider. A value of @c 0 would result in the trailing underline label getting all the available
+ width. A value of @c 1 would result in the leading underline label getting all the available width.
+ A value of @c .5 would result in each underline label getting 50% of the available width.
+ */
+@property(nonatomic, assign) CGFloat customUnderlineLabelDrawPriority;
+
+/**
+ When set to YES, the floating label floats when the view becomes the first responder. When
+ set to NO it disappears.
+
+ @note The default is YES.
+ */
+@property(nonatomic, assign) BOOL canFloatingLabelFloat;
+
+/**
+ This property toggles the error state (similar to @c isHighlighted, @c isEnabled, @c isSelected,
+ etc.) that is part of a general interpretation of the states outlined in the Material guidelines
+ for Text Fields. See the @c MDCContainedInputViewState enum for more information.
+ */
+@property(nonatomic, assign) BOOL isErrored;
+
+/**
+ This property toggles the activated state (similar to @c isHighlighted, @c isEnabled, @c
+ isSelected, etc.) that is part of a general interpretation of the states outlined in the Material
+ guidelines for Text Fields. See the @c MDCContainedInputViewState enum for more information.
+ */
+@property(nonatomic, assign) BOOL isActivated;
+
+/**
+  This method returns a color scheme for a given state.
+ */
+- (nonnull id<MDCContainedInputViewColorScheming>)containedInputViewColorSchemingForState:
+    (MDCContainedInputViewState)containedInputViewState;
```

Comment:
Not sure how this would work with the new state contrast ratio theming, since states will be affected by both color and elevation. 

It's traditionally the role of themers to determine how states are themed.  Moving it to be the responsibility of the component is contradicting the basic "dumb components" and "mutator themers" architecture of the library, which all other components follow.  

I'm concerned that deviating from the standard will introduce complexity in state theming.
State theming requires that themers have full access to all state-related component attributes and be able to choose values for any attribute in any state.

Reply:
I'm not sure either. FWIW, I haven't done anything with elevation at this point. I don't believe the material.io guidelines for text fields mention elevation, and I'm kind of using that as my source of truth for all this stuff.

Would calling it MDCContainedInputViewColorViewModel be better? We don't have to call it a scheme.

The rationale for having something like this in general: At any given point, the view needs to know what colors its subviews should be. Those colors change depending on what state the view is in. It needs to have some way to know and update the various colored elements when a state change occurs.

Some maybe good news:
- This stuff is in no way tied to MDCColorScheming. MDCColorSchemes are adapted into these schemes within the theming extensions. This may make changes to the color subsystem less difficult to deal with in the future.
- Clients will never have to mess with any of this stuff if they don't want to.

Reply:
InputChipView is my proposed solution for input chips. It is not the focus of this API review. With InputChipView, any UIView can be added as a chip. If a client would want material input chips to be displayed in the InputChipView, they could use MDCChipView and theme each chip they add as an input chip. 

The protocol intentionally has a lot of things going on. It's supposed to contain everything that these very similar input views with containers (like the textfield and inputchipview) have in common. That was the point. And it's not actually that big. To put it in perspective, MDCContainedInputView has 16 properties and methods. MDCTextInput has 21 properties and methods, and MDCTextInputController has 44.

And lastly, I think we as a team are moving away from themers. The new preference is theming extensions. Contained Input Views are intended to have theming extensions. Two already exist. They set appropriate MDCContainedInputViewColorSchemes for given states.